File: VoxelSniperCore/src/main/java/com/github/kevindagame/util/BrushOperation/BlockOperation.java
package com.github.kevindagame.util.BrushOperation;

import com.github.kevindagame.voxelsniper.blockdata.IBlockData;
import com.github.kevindagame.voxelsniper.location.VoxelLocation;
import com.github.kevindagame.voxelsniper.material.VoxelMaterial;

/**
 * Operation that is performed on a block. This operation modifies blockData.
 */
public class BlockOperation extends LocationOperation {

    private final IBlockData oldData;
    private IBlockData newData;

    public BlockOperation(VoxelLocation location, IBlockData oldData, IBlockData newData) {
        super(location);
        this.oldData = oldData;
        this.newData = newData;
    }

    public IBlockData getOldData() {
        return oldData;
    }

    public IBlockData getNewData() {
        return newData;
    }

    public void setNewData(IBlockData newData) {
        this.newData = newData;
    }

    /**
     *
     * @param newMaterial the material to create blockData for
     */
    public void setNewMaterial(VoxelMaterial newMaterial) {
        this.newData = newMaterial.createBlockData();
    }
}

File: VoxelSniperCore/src/main/java/com/github/kevindagame/util/brushOperation/BlockOperation.kt
package com.github.kevindagame.util.brushOperation

import com.github.kevindagame.snipe.Undo
import com.github.kevindagame.voxelsniper.blockdata.IBlockData
import com.github.kevindagame.voxelsniper.location.BaseLocation
import com.github.kevindagame.voxelsniper.material.VoxelMaterial

/**
 * Operation that is performed on a block. This operation modifies blockData.
 */
class BlockOperation(location: BaseLocation, val oldData: IBlockData, var newData: IBlockData) :
    BrushOperation(location) {
    private var applyPhysics = true

    constructor(location: BaseLocation, oldData: IBlockData, newData: IBlockData, applyPhysics: Boolean) : this(
        location,
        oldData,
        newData
    ) {
        this.applyPhysics = applyPhysics
    }

    fun applyPhysics(): Boolean {
        return applyPhysics
    }

    /**
     * @param newMaterial the material to create blockData for
     */
    fun setNewMaterial(newMaterial: VoxelMaterial) {
        newData = newMaterial.createBlockData()
    }

    override fun perform(undo: Undo): Boolean {
        val block = location.block
        undo.put(block)
        block.setBlockData(newData, applyPhysics())
        return false
    }
}